bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/SitePersistenceService.kt

```diff
@@ -0,0 +1,105 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ******************************************************************************/
+
+package org.eclipse.tractusx.bpdm.gate.service
+
+import org.eclipse.tractusx.bpdm.common.exception.BpdmNotFoundException
+import org.eclipse.tractusx.bpdm.common.util.replace
+import org.eclipse.tractusx.bpdm.gate.api.model.SiteGateInputRequest
+import org.eclipse.tractusx.bpdm.gate.entity.*
+import org.eclipse.tractusx.bpdm.gate.repository.GateAddressRepository
+import org.eclipse.tractusx.bpdm.gate.repository.LegalEntityRepository
+import org.eclipse.tractusx.bpdm.gate.repository.SiteRepository
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+
+@Service
+class SitePersistenceService(
+    private val siteRepository: SiteRepository,
+    private val legalEntityRepository: LegalEntityRepository,
+    private val addressRepository: GateAddressRepository
+) {
+
+    @Transactional
+    fun persistSitesBP(sites: Collection<SiteGateInputRequest>) {
+
+        //Finds Site in DB
+        val externalIdColl: MutableCollection<String> = mutableListOf()
+        sites.forEach { externalIdColl.add(it.externalId) }
+        val siteRecord = siteRepository.findByExternalIdIn(externalIdColl)
+
+        sites.forEach { site ->
+
+            val legalEntityRecord =
+                site.legalEntityExternalId.let {
+                    legalEntityRepository.findByExternalId(site.legalEntityExternalId) ?: throw BpdmNotFoundException("Business Partner", it)
+                }
+
+            val fullSite = site.toSiteGate(legalEntityRecord)
+
+            siteRecord.find { it.externalId == site.externalId }?.let { existingSite ->
+
+                val logisticAddressRecord =
+                    addressRepository.findByExternalId(site.externalId + "_site") ?: throw BpdmNotFoundException("Business Partner", "Error")
+
+                updateAddress(logisticAddressRecord, fullSite.mainAddress)
+
+                updateSite(existingSite, site, legalEntityRecord)
+
+                siteRepository.save(existingSite)
+            } ?: run {
+                siteRepository.save(fullSite)
+            }
+        }
+    }
+
+    private fun updateSite(site: Site, updatedSite: SiteGateInputRequest, legalEntityRecord: LegalEntity) {
+
+        site.bpn = updatedSite.bpn
+        site.name = updatedSite.site.name
+        site.externalId = updatedSite.externalId
+        site.legalEntity = legalEntityRecord
+        site.states.replace(updatedSite.site.states.map { toEntityAddress(it, site) })
+
+    }
+
+    private fun updateAddress(address: LogisticAddress, changeAddress: LogisticAddress) {
```

Comment:
Looks like duplicate code from address service to me. Use the address service update method instead